tidb-binlog/pump_client/selector_test.go

```diff
@@ -0,0 +1,28 @@
+package client
+
+import (
+	. "github.com/pingcap/check"
+)
+
+type testSelectorSuite struct{}
+
+var _ = Suite(&testSelectorSuite{})
+
+func (t *testSelectorSuite) TestHashTS(c *C) {
+	const (
+		nPumps = 11
+		nTS    = 1024
+	)
+	counter := [nPumps]int{}
+	for i := 0; i < nTS; i++ {
+		counter[hashTs(int64(i))%nPumps]++
```

Comment:
You need to cast the result of `hashTs` to an unsigned integer before taking modulo.

Currently `hashTs` may return a negative number, and `(-50) % 11 == -6`. The range of counter is thus -10 to 10, where the "0" slot is double-counted, i.e. the ideal distribution is like:

```
-10 = 4.5%
 -9 = 4.5%
...
 -1 = 4.5%
  0 = 9.1% <---
  1 = 4.5%
...
  9 = 4.5%
 10 = 4.5%
```

If we remove the double-counting, the deviation from average should be much less than 300.

Reply:
Yes, the problem is that in 32-bit environment (eg. go playground), `int(h.Sum32())` may be negative.

Reply:
You could do `uint(hashTs) % nPumps`